torch/jit/quantized.py

```diff
@@ -45,10 +50,200 @@ def extra_repr(self):
+        return ret
+
+
+class QuantizedLSTMCell(QuantizedRNNCellBase):
+    def __init__(self, other):
+        super(QuantizedLSTMCell, self).__init__(other)
+
+    @torch.jit.script_method
+    def forward(self, input, hx=None):
+        # type: (Tensor, Optional[Tuple[Tensor, Tensor]]) -> Tuple[Tensor, Tensor]
+        self.check_forward_input(input)
+        if hx is None:
+            zeros = torch.zeros(input.size(0), self.hidden_size, dtype=input.dtype, device=input.device)
+            _hx = (zeros, zeros)
+        else:
+            _hx = torch.jit._unwrap_optional(hx)
+        self.check_forward_hidden(input, _hx[0], '[0]')
+        self.check_forward_hidden(input, _hx[1], '[1]')
+        return _VF.quantized_lstm_cell(
+            input, _hx, self.weight_ih, self.weight_hh, self.bias_ih,
+            self.bias_hh, self.packed_ih, self.packed_hh, self.col_offsets_ih,
+            self.col_offsets_hh, self.scale_ih, self.scale_hh, self.zero_point_ih,
+            self.zero_point_hh
+        )
+
+
+class QuantizedGRUCell(QuantizedRNNCellBase):
+    def __init__(self, other):
+        super(QuantizedGRUCell, self).__init__(other)
+
+    @torch.jit.script_method
+    def forward(self, input, hx=None):
+        # type: (Tensor, Optional[Tensor]) -> Tensor
+        self.check_forward_input(input)
+        if hx is None:
+            _hx = torch.zeros(input.size(0), self.hidden_size, dtype=input.dtype, device=input.device)
+        else:
+            _hx = torch.jit._unwrap_optional(hx)
+        self.check_forward_hidden(input, _hx, '')
+        return _VF.quantized_gru_cell(
+            input, _hx, self.weight_ih, self.weight_hh, self.bias_ih,
+            self.bias_hh, self.packed_ih, self.packed_hh, self.col_offsets_ih,
+            self.col_offsets_hh, self.scale_ih, self.scale_hh, self.zero_point_ih,
+            self.zero_point_hh
+        )
+
+
+def quantize_rnn_cell_modules(module):
+    reassign = {}
+    for name, mod in module.named_modules():
+        if mod is module:
+            continue
+        new_mod = quantize_rnn_cell_modules(mod)
+        if new_mod != mod:
+            reassign[name] = new_mod
+    for name, mod in reassign.items():
+        setattr(module, name, mod)
+    if isinstance(module, torch.nn.LSTMCell):
+        return QuantizedLSTMCell(mod)
+    if isinstance(module, torch.nn.GRUCell):
+        return QuantizedGRUCell(mod)
+    if isinstance(module, torch.nn.RNNCell):
+        return QuantizedRNNCell(mod)
+
+    return module
+
+
 def quantize_linear_modules(module):
+    reassign = {}
     for name, mod in module.named_modules():
         if mod is module:
             continue
-        if isinstance(mod, torch.nn.Linear):
-            setattr(module, name, QuantizedLinear(mod))
-        quantize_linear_modules(mod)
+        new_mod = quantize_linear_modules(mod)
+        if new_mod != mod:
+            reassign[name] = new_mod
+
+    for name, mod in reassign.items():
+        setattr(module, name, mod)
```

Comment:
I don't understand why can't we do the reassignments directly in the loop. What's the deal there?

Reply:
Not exactly sure why, but things seem to break when I do the mutation in the loop. I seem to remember there was an exception thrown before, but now when i've tried it with the assignment in the loop, it seems to screw up the contents of the dictionary somehow